kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/RecordEncryptionFilter.java

```diff
@@ -114,13 +115,18 @@ private CompletionStage<ProduceRequestData> maybeEncodeProduce(ProduceRequestDat
                     if (!unresolvedTopicNames.isEmpty() && unresolvedKeyPolicy == UnresolvedKeyPolicy.REJECT) {
                         return CompletableFuture.failedFuture(new UnresolvedKeyException("failed to resolve key for: " + unresolvedTopicNames));
                     }
+
+                    // metrics generation for not encrypted record
+                    generateMetrics(context.getVirtualClusterName(), unresolvedTopicNames, topicNameToData);
+
                     var futures = kekSelection.topicNameToKekId().entrySet().stream().flatMap(e -> {
                         String topicName = e.getKey();
                         var kekId = e.getValue();
                         TopicProduceData tpd = topicNameToData.get(topicName);
                         return tpd.partitionData().stream().map(ppd -> {
                             MemoryRecords records = (MemoryRecords) ppd.records();
                             return encryptionManager.encrypt(
+                                    context.getVirtualClusterName(),
```

Comment:
why did you choose to give the responsibility for ticking the metric to the EncryptionManager?    You want as fewer parts of the system as possible to know about Metrics - think Parnas and Information Hiding.

So you've already given REF knowledge for the metrics for one path (fine).. so now strive to keep to give it the responsibility for recording metrics for the other path too.

Hint - you'll want to chain onto EncryptionManager.encrypt() future :)